DIPs/DIP1xxx-rval_to_ref.md

```diff
@@ -0,0 +1,345 @@
+
+A related issue is with relation to generic code which reflects or received a function by alias. Such generic code may want to call that function, but it is often the case that details about the ref-ness of arguments lead to incorrect semantic expressions in the generic code depending on the arguments, necessitating additional compile-time logic to identify the ref-ness of function arguments and implement appropriate workarounds on these conditions. This leads to longer, more brittle, and less-maintainable generic code. It is also much harder to write correctly the first time, and such issues may only emerge in niche use cases at a later time.
+
+With these cases in mind, the existing rule feels out-dated or inappropriate, and the presence of the rule may often lead to aggravation while trying to write simple, readable code.  
+Calling a function should be simple and orthogonal, generic code should not have to concern itself with details about ref-ness of function parameters, and users should not be required to jump through hoops when `ref` appears in API's they encounter.
+
+Consider the example:
+```d
+void fun(int x);
+
+fun(10); // <-- this is how simple calling a function should be
+```
+But when `ref` is involved:
+```d
+void fun(ref int x);
+
+fun(10); // <-- compile error; not an lvalue!!
+```
+Necessitating the workaround:
+```d
+int temp = 10;
+fun(temp);
+```
+In practise, the argument would likely be some larger struct type rather than 'int', but the inconvenience applies generally.
+
+This inconvenience also extends more broadly to cases including:
+```d
+fun(10);       // literals
+fun(gun());    // return values from functions
+fun(x.prop);   // properties
+fun(x + y);    // expressions
+fun(my_short); // implicit type conversions (ie, short->int promotion)
+// etc... (basically, most things you pass to functions)
+```
+The work-around can bloat the number of lines around the call-site significantly, and the user needs to declare names for all the temporaries, polluting the local namespace, and often for expressions where no meaningful name exists, leading to.
+
+The generic case may appear in a form like this:
+```d
+void someMeta(alias userFun)()
+{
+    userFun(getValue());
+}
+
+void fun(int x);
+void gun(ref const(int) x);
+
+unittest
+{
+    someMeta!(fun)(); // no problem
+    someMeta!(gun)(); // oh no, can't receive rvalue!
+}
+```
+Necessitating a workaround that may look like:
+```d
+void someMeta(alias userFun)()
+{
+    std.algorithm : canFind;
+    static if(canFind(__traits(getParameterStorageClasses, userFun, 0), "ref"))
+    {
+        auto x = getValue();
+        userFun(x);
+    }
+    else
+    {
+        userFun(getValue());
+    }
+}
+```
+This example situation is simplified, but it is often that such issues appear in complex aggregate meta, which may be difficult to understand, or the issue is caused indirectly at some layer the user did not author.
```

Comment:
Be careful about authoritative claims about frequency, or side effects that are hard to quantify. If you have precise claims to make, with demonstrative data to back them up, then the data need to be referenced somehow. Otherwise, consider language such as "can lead to" or "may cause", with examples to demonstrate. This applies not just here with these lines, but throughout the document.